### PR TITLE
Shutdown hooks

### PR DESCRIPTION
**Pull request by a-palchikov**
This PR removes implicit reference to magnet (for `magnet.Shutdown` invocations) and instead embeds the shutdown hooks API into the template.
It is based on the latest upstream [master](https://github.com/magefile/mage/commit/4cf3cfcc82bab35b01ab1657c2aa2d0faeadac49) - I will rebase it once/if https://github.com/knisbet/mage/pull/1 is merged.

@knisbet PTAL.

### CODE REVIEW

**Review by knisbet — APPROVED** [review]
So my intention was always to reach out to the upstream authors to get this properly included if we continued this project... with this cleaner implementation it might be worth reaching out.

**Comment by a-palchikov** [comment]
I submitted a [PR](https://github.com/magefile/mage/pull/345) upstream. Let me know if you can merge the upstream bump and this eventually.

**Comment by knisbet** [comment]
Sure, just did.

File: parse/parse_test.go [hunk 1]
```diff
@@ -71,6 +71,20 @@ func TestParse(t *testing.T) {
 		t.Fatalf("expected DefaultName to be ReturnsNilError")
 	}
 
+	if info.DeinitFunc == nil {
```
Comment (knisbet) [inline parse/parse_test.go:74]: Since Deinit is the less common case, probably want to test where it's not universally set?

Comment (a-palchikov) [reply parse/parse_test.go:74]: Not sure I fully understood your concern but I added two more test-cases for Deinit:
  * ensure that `DeinitFunc` is not initialized by default
  * ensure that it is properly scheduled for targets if defined

Comment (knisbet) [reply parse/parse_test.go:74]: I mean we're the only user of deinit, but by adding to parse_test this mostly tests when deinit is set, not when it isn't set.

Comment (a-palchikov) [reply parse/parse_test.go:74]: https://github.com/magefile/mage/pull/345/files#diff-daea2a9086a9f20e21f1ee99180680de0b4f12ce020fd4c99914b466db5a2c9dR130 tests that Deinit is not set by default - is that what you meant?

**Pull request merged**